### PR TITLE
fix #231 allow square brackets in path

### LINKED ISSUE #220: Cannot clone repo on windows
When trying to clone this repository on windows I get the following error:

```
kamik@T460p MINGW64 /d/sandbox
$ git clone https://github.com/webpack-contrib/copy-webpack-plugin
Cloning into 'copy-webpack-plugin'...
remote: Counting objects: 796, done.
remote: Compressing objects: 100% (34/34), done.
remote: Total 796 (delta 18), reused 31 (delta 13), pack-reused 748
Receiving objects: 100% (796/796), 266.90 KiB | 0 bytes/s, done.
Resolving deltas: 100% (440/440), done.
fatal: cannot create directory at 'tests/helpers/[special?directory]': Invalid argument
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry the checkout with 'git checkout -f HEAD'
```

I think the real error is:
```
fatal: cannot create directory at 'tests/helpers/[special?directory]': Invalid argument
```

The special character `?` is not permitted in windows directories.

### LINKED ISSUE #231: Won't copy files when there are square brackets [ ] in the path
A tiny repo to reproduce the issue:
https://github.com/flawiddsouza/Reproduce-bracket-bug-in-copy-webpack-plugin

Go to each folder & run:
npm install
npm start

The folder without brackets will copy the file "from" to "to" but the one with brackets will not.

### PR DESCRIPTION
**Pull request by loveky**
fixes #231 fixes #220

### CODE REVIEW

**Comment by loveky** [comment]
@evilebottnawi 

I have a problem when setting up development environment on Windows. Because the file name/dir name under `tests/helpers`  containes `?` `*` which is not allow on Windows. git clone/git reset just fails.

image


Currently what I have done:

1.  clone the reproduce repo in #231; npm install; modify `copy-webpack-plugin` under node_modules, verify it solves the issue.
2. on mac clone `copy-webpack-plugin` repo. apply the same change, run test to make sure it didn't broken on mac system.

any suggestion?

**Comment by evilebottnawi** [comment]
@loveky can you fix it in this PR? Just disable this test if you in windows

**Comment by loveky** [comment]
got it. will work on that

File: src/utils/escape.js [hunk 1]
```diff
@@ -8,7 +8,8 @@ export default function escape(context, from) {
         // Handles special characters in paths
         const absoluteContext = path.resolve(context)
             .replace(/\\/, '/')
-            .replace(/[\*|\?|\!|\(|\)|\[|\]|\{|\}]/g, (substring) => `\\${substring}`);
+            .replace(/[\*|\?|\!|\(|\)|\{|\}]/g, (substring) => `\\${substring}`)
+            .replace(/\[/g, '[[]');
```
Comment (evilebottnawi) [inline src/utils/escape.js]: Looks strange. Can you describe why past solution doesn't work?

File: scripts/createSpecialDirectory.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    '[special?directory]/(special-*file).txt': 'special',
+    '[special?directory]/directoryfile.txt': 'new'
+};
+
+Object.keys(specialFiles).forEach(function (originFile) {
+    const file = removeIllegalCharacterForWindows(originFile);
+    const dir = path.dirname(file);
+    mkdirp.sync(path.join(baseDir, dir));
+    fs.writeFileSync(path.join(baseDir, file), specialFiles[originFile]);
+});
```
[the remark this answers is not in the document]
Comment (loveky) [reply scripts/createSpecialDirectory.js]: which file should I add `\n`? `createSpecialDirectory.js` itself? or the files created by it? 🤣

[the remark this answers is not in the document]
Comment (evilebottnawi) [reply scripts/createSpecialDirectory.js]: @loveky in this file :smile:

[the remark this answers is not in the document]
Comment (loveky) [reply scripts/createSpecialDirectory.js]: done 😃

File: tests/index.js [hunk 3]
```diff
@@ -15,6 +15,8 @@ import cacache from 'cacache';
 import isGzip from 'is-gzip';
 import zlib from 'zlib';
 
+import removeIllegalCharacterForWindows from '../scripts/removeIllegalCharacterForWindows';
```
Comment (filipesilva) [inline tests/index.js]: This should probably be in `src/utils` instead of `scripts/` otherwise it won't get published.

**Comment by filipesilva** [comment]
Tried a build from this PR on my windows machine with a directory using parenthesis and the assets were copied.

**Comment by filipesilva** [comment]
The lack of new tests might lead to regressions on special character support though.

**Comment by loveky** [comment]
Hi @filipesilva , the square brackets is already in the [tests](https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/tests/index.js#L268-L287). Should I add more tests? any suggestion?

**Comment by filipesilva** [comment]
I did not realize that, sorry for the confusion!

**Comment by evilebottnawi** [comment]
@filipesilva can we merge this, for me LGTM, but i don't have windows near

File: tests/index.js (hunk 3 above)
Comment (evilebottnawi) [reply tests/index.js]: @loveky yep, move this to `tests/utils`

**Comment by filipesilva** [comment]
I can say that in my windows machine the test suite in `npm test` passed. Would you like me to check something else on my machine?

**Comment by evilebottnawi** [comment]
@filipesilva not, just this PR

File: tests/index.js [hunk 4]
```diff
@@ -984,7 +993,7 @@ describe('apply function', () => {
                     'nested/nestedfile.txt'
                 ],
                 patterns: [{
-                    from: '[special?directory]'
+                    from: (path.sep === '/' ? '[special?directory]' : '[specialdirectory]')
```
[the remark this answers is not in the document]
Comment (evilebottnawi) [reply tests/index.js]: @clydin we already use `appveyor` for many plugins, need merge `webpack-default` here, will be done in near future

File: scripts/createSpecialDirectory.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+const path = require('path');
+const fs = require('fs');
+const removeIllegalCharacterForWindows = require('./removeIllegalCharacterForWindows');
+
+const baseDir = 'compiled_tests/helpers';
+
+const specialFiles = {
+    '[special?directory]/nested/nestedfile.txt': '',
+    '[special?directory]/(special-*file).txt': 'special',
+    '[special?directory]/directoryfile.txt': 'new'
```
[the remark this answers is not in the document]
Comment (evilebottnawi) [reply scripts/createSpecialDirectory.js]: @loveky yep, add test

**Comment by evilebottnawi** [comment]
wait @filipesilva approve

**Comment by filipesilva** [comment]
Oh sorry, didn't mean to keep you waiting!

**Pull request merged**